File: challenge/leet.gcc/easy/q350/q350.cpp
/**
 * 350. Intersection of Two Arrays II
 * https://leetcode.com/problems/intersection-of-two-arrays-ii/
 */
#include <vector>
#include <iostream>
#include <algorithm>
#include <numeric>
#include <string>

using namespace std;

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());

        vector<int> res;
        for (auto i = nums1.begin(), j = nums2.begin(); i != nums1.end() and j != nums2.end(); ) {
            if (*i < *j) i++;
            else if (*i > *j) j++;
            else {
                res.push_back(*i);
                i++, j++;
            }
        }
        return res;
    }
};

class Solution2 {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());

        vector<int> res;
        int i = 0, s1 = nums1.size(), j = 0, s2 = nums2.size();
        while (i < s1 and j < s2) {
            if (nums1[i] < nums2[j]) i++;
            else if (nums1[i] > nums2[j]) j++;
            else {
                res.push_back(nums1[i]);
                i++, j++;
            }
        }
        return res;
    }
};


string reduce(vector<int> v) {
    return string("[") + (v.size() > 0 ? reduce(next(v.begin()), v.end(), to_string(v[0]),
        [](string a, int b) {return move(a) + ", " + to_string(b); }) : "") + string("]");
}

int main() {
    Solution2 s;
    vector<int> n1;
    vector<int> n2;

    n1 = vector<int>{ 1,2,2,1 };
    n2 = vector<int>{ 1,2 };
    cout << "[1, 2] ?" << endl
        << reduce(s.intersect(n1, n2)) << endl;

    n1 = vector<int>{ 1,2 };
    n2 = vector<int>{ 1,2,2,1 };
    cout << "[1, 2] ?" << endl
        << reduce(s.intersect(n1, n2)) << endl;

}

File: challenge/leet/easy/q350.py
'''
350. Intersection of Two Arrays II
https://leetcode.com/problems/intersection-of-two-arrays-ii/
'''
import unittest
from collections import defaultdict
from typing import List


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        bits = defaultdict(lambda : 0)
        for n in nums2:
            bits.update({n: bits[n] + 1})

        ans = []
        for m in nums1:
            if bits[m] > 0:
                ans.append(m)
                bits.update({m: bits[m] - 1})
        return ans


class Solution2:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        s1 = defaultdict(lambda : 0)
        for n in nums1:
            s1.update({n: s1[n] + 1})

        s2 = defaultdict(lambda : 0)
        for n in nums2:
            s2.update({n: s2[n] + 1})

        ans = []
        for m in s1:
            if m in s2:
                ans.extend([m] * min(s1[m], s2[m]))
        return ans


class MyTestCase(unittest.TestCase):
    def test_something(self):
        s = Solution2()
        self.assertEqual([1,2], s.intersect([1,2,2,1], [1,2]))
        self.assertEqual([2,2], s.intersect([1,2,2,1], [2,2]))
        self.assertEqual([4,9], s.intersect([4,9,5], [9,4,9,8,4]))


if __name__ == '__main__':
    unittest.main()
